File: Firebase/ios/Bonfire/Mobile/Users/FirebaseUserDatabase.swift
import Foundation
import RxSwift
import Firebase

class FirebaseUserDatabase: UserDatabase {
    let usersDB: FIRDatabaseReference

    init(usersDB: FIRDatabaseReference) {
        self.usersDB = usersDB
    }

    func observeUsers() -> Observable<Users> {
        return usersDB.rx_readValue().map(toUsers)
    }

    func readUserFrom(userID: String) -> Observable<User> {
        return usersDB.child(userID).rx_readOnce().map(User.init)
    }

    func writeCurrentUser(user: User) {
        usersDB.child(user.id).setValue(user.asFirebaseValue())
    }

    private func toUsers(snapshot: FIRDataSnapshot) throws -> Users {
        let users = try snapshot.children.allObjects.map({try User(snapshot: $0 as! FIRDataSnapshot)})
        return Users(users: users)
    }
}

File: Firebase/android/app/src/main/java/com/novoda/bonfire/user/database/FirebaseUserDatabase.java
package com.novoda.bonfire.user.database;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.novoda.bonfire.rx.OnSubscribeDatabaseListener;
import com.novoda.bonfire.user.data.model.User;
import com.novoda.bonfire.user.data.model.Users;

import java.util.ArrayList;
import java.util.List;

import rx.Observable;
import rx.functions.Func1;

import static com.novoda.bonfire.rx.RxSingleValueListener.listenToSingleValueEvents;

public class FirebaseUserDatabase implements UserDatabase {

    private final DatabaseReference usersDB;

    public FirebaseUserDatabase(FirebaseDatabase firebaseDatabase) {
        usersDB = firebaseDatabase.getReference("users");
    }

    @Override
    public Observable<Users> observeUsers() {
        return Observable.create(new OnSubscribeDatabaseListener<>(usersDB, toUsers()));
    }

    @Override
    public Observable<User> readUserFrom(String userId) {
        return listenToSingleValueEvents(usersDB.child(userId), as(User.class));
    }

    @Override
    public void writeCurrentUser(User user) {
        usersDB.child(user.getId()).setValue(user); //TODO handle errors
    }

    private Func1<DataSnapshot, Users> toUsers() {
        return new Func1<DataSnapshot, Users>() {
            @Override
            public Users call(DataSnapshot dataSnapshot) {
                Iterable<DataSnapshot> children = dataSnapshot.getChildren();
                List<User> users = new ArrayList<>();
                for (DataSnapshot child : children) {
                    User message = child.getValue(User.class);
                    users.add(message);
                }
                return new Users(users);
            }
        };
    }

    private <T> Func1<DataSnapshot, T> as(final Class<T> tClass) {
        return new Func1<DataSnapshot, T>() {
            @Override
            public T call(DataSnapshot dataSnapshot) {
                return dataSnapshot.getValue(tClass);
            }
        };
    }
}
